Updates to collision model google test + some doc

amcastro-tri opened the pull request:
This PR simply:
1. Updates documentation for `Model::collidingPoints` and `Model::collidingPointsCheckOnly`
2. Updates `model_test.cc` to use `EXPECT_NEAR` and `isApprox`, which are more appropriate for floating point comparisons.
3. Demonstrates how to use `AngleAxis` to setup simple isometries in `model_test.cc`.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/2402) &emsp; Multiple assignees:&emsp;<img alt="@jwnimmer-tri" height="20" width="20" align="absmiddle" src="https://avatars.githubusercontent.com/u/17596505?s=40&v=3">&nbsp;<a href="/robotlocomotion/drake/pulls/assigned/jwnimmer-tri">jwnimmer-tri</a>,&emsp;<img alt="@sherm1" height="20" width="20" align="absmiddle" src="https://avatars.githubusercontent.com/u/4088016?s=40&v=3">&nbsp;<a href="/robotlocomotion/drake/pulls/assigned/sherm1">sherm1</a>
<!-- Reviewable:end -->

jwnimmer-tri commented:
+@jwnimmer-tri for platform review.

<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> #### Updates to collision model google test \+ some doc
> 
> This PR simply:
> 
> 1. Updates documentation for `Model::collidingPoints` and `Model::collidingPointsCheckOnly`
> 
> 2. Updates `model_test.cc` to use `EXPECT_NEAR` and `isApprox`, which are more appropriate for floating point comparisons.
> 
> 3. Demonstrates how to use `AngleAxis` to setup simple isometries in `model_test.cc`.
> 
> 
> 
</details>


---

Review status: 0 of 3 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2402#-:-KI8_iLQWmunh9Gs3zdG:-1800536653)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
+@sherm1 for feature review.
Special thanks to @david-german-tri who helped me find out about `EXPECT_NEAR`!

jwnimmer-tri commented:
*[drake/systems/plants/collision/test/model_test.cc, line 84 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2402#-KI8gOYU7ipwYVqmP46x:-KI8gOYU7ipwYVqmP46y:-530320803) ([raw file](https://github.com/robotlocomotion/drake/blob/7547a380d5ebb2c1d26fd50204f37116abe7ebff/drake/systems/plants/collision/test/model_test.cc#L84)):*
> ```cpp
>   EXPECT_NEAR(1.0, points[0].getDistance(), tolerance);
>   // Normal is on body B.
>   EXPECT_TRUE(points[0].getNormal().isApprox(Vector3d(-1, 0, 0)));
> ```

@liangfok Could you please comment on whether using eigen's relative-tolerance `isApprox` is appropriate in unit tests (such as on this line, and the similar ones below), or if the matrix-comparison helper for Drake must be used instead.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2402)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Reviewed 3 of 3 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/systems/plants/collision/test/model_test.cc, line 84 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2402#-KI8gOYU7ipwYVqmP46x:-KI8i1IJkPBz8Aspiu2r:181865817) ([raw file](https://github.com/robotlocomotion/drake/blob/7547a380d5ebb2c1d26fd50204f37116abe7ebff/drake/systems/plants/collision/test/model_test.cc#L84)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> @liangfok Could you please comment on whether using eigen's relative-tolerance `isApprox` is appropriate in unit tests (such as on this line, and the similar ones below), or if the matrix-comparison helper for Drake must be used instead.
</details>
Drake's matrix comparison helper, which is defined in `eigen_matrix_compare.h` is more robust in terms of handling corner cases like `NaN` and infinity values. It is also more expressive in terms of being able to handle tolerance values that are relative to the size of the values being compared. 

In this case, perhaps @amcastro-tri can make the best judgement call regarding whether the values being compared here can potentially be `NaN`, infinity, or so huge that absolute tolerance may result in a bad comparision.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2402)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
+1 for cleaning up this mess!

<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> +@sherm1 for feature review.
> 
> Special thanks to @david-german-tri who helped me find out about `EXPECT_NEAR`!
</details>


---

Reviewed 3 of 3 files at r1.
Review status: 2 of 3 files reviewed at latest revision, 8 unresolved discussions.

---

*[drake/systems/plants/collision/Model.h, line 132 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2402#-KI8grcadCmaHPGZHeZV:-KI8grcadCmaHPGZHeZW:1584028878) ([raw file](https://github.com/robotlocomotion/drake/blob/7547a380d5ebb2c1d26fd50204f37116abe7ebff/drake/systems/plants/collision/Model.h#L132)):*
> ```objc
>   }
> 
>   // @brief Given a vector of points in world coordinates, returns the indices
> ```

This used to be a doxygen comment but now it is not. Was that intentional? 

Also, you don't need to say `@brief` any more since PR #2359.

---

*[drake/systems/plants/collision/Model.h, line 139 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2402#-KI8h0DBDV5I-WWGKHBA:-KI8h0DBDV5I-WWGKHBB:-1805487455) ([raw file](https://github.com/robotlocomotion/drake/blob/7547a380d5ebb2c1d26fd50204f37116abe7ebff/drake/systems/plants/collision/Model.h#L139)):*
> ```objc
>   // collision_threshold located at input_points[i] collides with any part of
>   // the model. The result is returned as a vector of indexes in input_points
>   // that do collide with the model.
> ```

I'm guessing that the points are not checked against one another but only point-to-existing-geometry. It would be good to clarify that in the comment either way.

---

*[drake/systems/plants/collision/Model.h, line 161 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2402#-KI8iKGxclrt46qLnWxx:-KI8iKGxclrt46qLnWxy:1972559848) ([raw file](https://github.com/robotlocomotion/drake/blob/7547a380d5ebb2c1d26fd50204f37116abe7ebff/drake/systems/plants/collision/Model.h#L161)):*
> ```objc
>   // collision_threshold located at input_points[i] collides with any part of
>   // the model. This method returns as soon as any of these spheres collides
>   // with the model.
> ```

Same comments as above.

---

*[drake/systems/plants/collision/test/model_test.cc, line 53 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2402#-KI8kIvYRLl_R7ymPQ51:-KI8kIvYRLl_R7ymPQ52:-1025921707) ([raw file](https://github.com/robotlocomotion/drake/blob/7547a380d5ebb2c1d26fd50204f37116abe7ebff/drake/systems/plants/collision/test/model_test.cc#L53)):*
> ```cpp
>       AngleAxisd(M_PI_2,Vector3d::UnitZ()).toRotationMatrix();
> 
>   const double tolerance = 1.0e-6;
> ```

What is the justification for this hard coded and extremely loose tolerance? Is that because we're asking the collision model for an approximate solution? I would like to see an explanation here and ideally I'd like to see this tolerance being part of the request to the collision model.

---

*[drake/systems/plants/collision/test/model_test.cc, line 84 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2402#-KI8gOYU7ipwYVqmP46x:-KI8jjd6uohqBPn0MXEp:1807419721) ([raw file](https://github.com/robotlocomotion/drake/blob/7547a380d5ebb2c1d26fd50204f37116abe7ebff/drake/systems/plants/collision/test/model_test.cc#L84)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Drake's matrix comparison helper, which is defined in `eigen_matrix_compare.h` is more robust in terms of handling corner cases like `NaN` and infinity values. It is also more expressive in terms of being able to handle tolerance values that are relative to the size of the values being compared. 
> 
> In this case, perhaps @amcastro-tri can make the best judgement call regarding whether the values being compared here can potentially be `NaN`, infinity, or so huge that absolute tolerance may result in a bad comparision.
</details>
Why was this changed from an exact comparison to an approximate one? Integers are represented exactly in floating point -- is there some approximation being done here? BTW, Eigen is using 1e-12 as the tolerance here.

---

*[drake/systems/plants/collision/test/model_test.cc, line 86 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2402#-KI8lLIPLK7-9aRrnvub:-KI8lLIPLK7-9aRrnvuc:-777611155) ([raw file](https://github.com/robotlocomotion/drake/blob/7547a380d5ebb2c1d26fd50204f37116abe7ebff/drake/systems/plants/collision/test/model_test.cc#L86)):*
> ```cpp
>   EXPECT_TRUE(points[0].getNormal().isApprox(Vector3d(-1, 0, 0)));
>   EXPECT_TRUE(points[0].getPtA().isApprox(Vector3d(0.5, 0, 0)));
>   EXPECT_TRUE(points[0].getPtB().isApprox(Vector3d(0.5, 0, 0)));
> ```

Same here -- if these aren't exact results something is wrong. Both 0 and 0.5 are exactly representable in binary floating point.

---

*[drake/systems/plants/collision/test/model_test.cc, line 98 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2402#-KI8laMitI3UovrSdZh8:-KI8laMitI3UovrSdZh9:-526048411) ([raw file](https://github.com/robotlocomotion/drake/blob/7547a380d5ebb2c1d26fd50204f37116abe7ebff/drake/systems/plants/collision/test/model_test.cc#L98)):*
> ```cpp
>       Vector3d(0.5, 0.5, 0)));
>   EXPECT_TRUE(points[1].getPtB().isApprox(
>       Vector3d(-sqrt(2) / 4, sqrt(2) / 4, 0)));
> ```

OK, these can be approximate (should be accurate to around 1e-14, so this provides two orders of magnitude of slop). But actually I don't understand why we're expecting such a good solution here if we were only expecting 1e-6 earlier (that's a million times sloppier than 1e-12).

---

*[drake/systems/plants/collision/test/model_test.cc, line 107 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2402#-KI8mA6g_PWRO8CrVT_m:-KI8mA6g_PWRO8CrVT_n:-758158901) ([raw file](https://github.com/robotlocomotion/drake/blob/7547a380d5ebb2c1d26fd50204f37116abe7ebff/drake/systems/plants/collision/test/model_test.cc#L107)):*
> ```cpp
>   EXPECT_TRUE(points[2].getNormal().isApprox(Vector3d(0, -1, 0)));
>   EXPECT_TRUE(points[2].getPtA().isApprox(Vector3d(1, 0.5, 0)));
>   EXPECT_TRUE(points[2].getPtB().isApprox(Vector3d(-0.5, 0, 0)));
> ```

Shouldn't these last three be exact?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2402#-:-KI8mG7RlQqynfIU7g6u:-201771471)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Review status: 2 of 3 files reviewed at latest revision, 9 unresolved discussions.

---

*[drake/systems/plants/collision/test/model_test.cc, line 93 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2402#-KI8qDt_ccoMoscF6El-:-KI8qDtax19g0-vW21yX:1707759772) ([raw file](https://github.com/robotlocomotion/drake/blob/4f39597991986b22e089a970a69468385f21eaaf/drake/systems/plants/collision/test/model_test.cc#L93)):*
> ```cpp
>   EXPECT_EQ(id1, points[1].getIdA());
>   EXPECT_EQ(id3, points[1].getIdB());
>   EXPECT_NEAR(1.6213203435596428, points[1].getDistance(), tolerance);
> ```

Better not to have magic constants if possible. I fed this one to one of my long-time [favorite web sites](http://isc.carma.newcastle.edu.au) -- the "inverse symbolic calculator" (you give it the answer it magically tells you how it was computed!). This appears to be (3*sqrt(2)-1)/2.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2402)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 3 of 3 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[drake/systems/plants/collision/bullet_model.h, line 117 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2402#-KI8xq-3b3MyprfZlwFj:-KI8xq-3b3MyprfZlwFk:-430432224) ([raw file](https://github.com/robotlocomotion/drake/blob/7547a380d5ebb2c1d26fd50204f37116abe7ebff/drake/systems/plants/collision/bullet_model.h#L117)):*
> ```objc
> 
>   bool collidingPointsCheckOnly(
>       const std::vector<Eigen::Vector3d>& input_points,
> ```

It is (in general) defective for the h file param name and cc file param name to be different.  It looks like this only changed the h file param name.

---

*[drake/systems/plants/collision/bullet_model.h, line 121 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2402#-KI8xxM-0mKrKbtCOKlC:-KI8xxM-0mKrKbtCOKlD:-430432224) ([raw file](https://github.com/robotlocomotion/drake/blob/7547a380d5ebb2c1d26fd50204f37116abe7ebff/drake/systems/plants/collision/bullet_model.h#L121)):*
> ```objc
> 
>   std::vector<size_t> collidingPoints(
>       const std::vector<Eigen::Vector3d>& input_points,
> ```

It is (in general) defective for the h file param name and cc file param name to be different.  It looks like this only changed the h file param name.

---

*[drake/systems/plants/collision/Model.h, line 151 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2402#-KI8zhmetSEC8IeYk7SU:-KI8zhmfNMuXas6a-WKR:-1625350535) ([raw file](https://github.com/robotlocomotion/drake/blob/4f39597991986b22e089a970a69468385f21eaaf/drake/systems/plants/collision/Model.h#L151)):*
> ```objc
>   // @see systems/plants/test/collidingPointsTest.m for a Matlab test.
>   virtual std::vector<size_t> collidingPoints(
>       const std::vector<Eigen::Vector3d>& points, double collision_threshold) {
> ```

The "points" parameter here is called "input_points" in the documentation; I assume the parameter name was supposed to change also.

If that is true, then this would also then apply: It is (in general) defective for the h file param name and cc file param name to be different.  It looks like this only changed the h file param name.

---

*[drake/systems/plants/collision/Model.h, line 171 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2402#-KI8zwsj4vEy8Qb9hTdP:-KI8zwskvPg3JWnnNUvr:-430432224) ([raw file](https://github.com/robotlocomotion/drake/blob/4f39597991986b22e089a970a69468385f21eaaf/drake/systems/plants/collision/Model.h#L171)):*
> ```objc
>   // `false` otherwise.
>   virtual bool collidingPointsCheckOnly(
>       const std::vector<Eigen::Vector3d>& input_points,
> ```

It is (in general) defective for the h file param name and cc file param name to be different.  It looks like this only changed the h file param name.

---

*[drake/systems/plants/collision/test/model_test.cc, line 84 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2402#-KI8gOYU7ipwYVqmP46x:-KI9-padLKvJlM7OzhjM:2053183550) ([raw file](https://github.com/robotlocomotion/drake/blob/7547a380d5ebb2c1d26fd50204f37116abe7ebff/drake/systems/plants/collision/test/model_test.cc#L84)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> Why was this changed from an exact comparison to an approximate one? Integers are represented exactly in floating point -- is there some approximation being done here? BTW, Eigen is using 1e-12 as the tolerance here.
</details>
I am going to reviewable-resolve myself out of this discussion.

However, just to make sure I understand: the tolerance 1e-6 above used for EXPECT_NEAR is an _absolute_ tolerance, and the Eigen 1e-12 is a _proportional_ tolerance, correct?  Some of the downthread comments make me think this isn't 100% clear in the discussions.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2402)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 1 of 4 files reviewed at latest revision, 13 unresolved discussions.

---

*[drake/systems/plants/collision/Model.h, line 132 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2402#-KI8grcadCmaHPGZHeZV:-KI8qo8f2ljpZmsO60tw:-1619647576) ([raw file](https://github.com/robotlocomotion/drake/blob/7547a380d5ebb2c1d26fd50204f37116abe7ebff/drake/systems/plants/collision/Model.h#L132)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> This used to be a doxygen comment but now it is not. Was that intentional? 
> 
> Also, you don't need to say `@brief` any more since PR #2359.
</details>
It IS meant to be a doxygen comment! I don't know what I was thinking. Thanks!
also removed the unnecessary @brief

---

*[drake/systems/plants/collision/Model.h, line 139 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2402#-KI8h0DBDV5I-WWGKHBA:-KI8r6b-wCRreCOoEM2f:676635175) ([raw file](https://github.com/robotlocomotion/drake/blob/7547a380d5ebb2c1d26fd50204f37116abe7ebff/drake/systems/plants/collision/Model.h#L139)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> I'm guessing that the points are not checked against one another but only point-to-existing-geometry. It would be good to clarify that in the comment either way.
</details>
Done

---

*[drake/systems/plants/collision/Model.h, line 161 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2402#-KI8iKGxclrt46qLnWxx:-KI92cJj3wbIcjiMnmxX:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/7547a380d5ebb2c1d26fd50204f37116abe7ebff/drake/systems/plants/collision/Model.h#L161)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> Same comments as above.
</details>
Done.

---

*[drake/systems/plants/collision/Model.h, line 151 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2402#-KI8zhmetSEC8IeYk7SU:-KI92V6bgj6GyAvcvVAJ:-497468049) ([raw file](https://github.com/robotlocomotion/drake/blob/4f39597991986b22e089a970a69468385f21eaaf/drake/systems/plants/collision/Model.h#L151)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> The "points" parameter here is called "input_points" in the documentation; I assume the parameter name was supposed to change also.
> 
> If that is true, then this would also then apply: It is (in general) defective for the h file param name and cc file param name to be different.  It looks like this only changed the h file param name.
</details>
Fixed

---

*[drake/systems/plants/collision/Model.h, line 171 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2402#-KI8zwsj4vEy8Qb9hTdP:-KI92i6MQGLErRsgWiGK:-497468049) ([raw file](https://github.com/robotlocomotion/drake/blob/4f39597991986b22e089a970a69468385f21eaaf/drake/systems/plants/collision/Model.h#L171)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> It is (in general) defective for the h file param name and cc file param name to be different.  It looks like this only changed the h file param name.
</details>
Fixed

---

*[drake/systems/plants/collision/test/model_test.cc, line 53 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2402#-KI8kIvYRLl_R7ymPQ51:-KI8rwi3K1TV0m4glki3:-150302074) ([raw file](https://github.com/robotlocomotion/drake/blob/7547a380d5ebb2c1d26fd50204f37116abe7ebff/drake/systems/plants/collision/test/model_test.cc#L53)):*
> ```cpp
>       AngleAxisd(M_PI_2,Vector3d::UnitZ()).toRotationMatrix();
> 
>   const double tolerance = 1.0e-6;
> ```

From the tests that I've run so far is not uncommon to get results from Bullet within 1e-9. However that changes with the complexity of the geometry. For this simple case the tests are passing with a precision of `epsilon`. I updated to use `epsilon` instead. 

>  I'd like to see this tolerance being part of the request to the collision model.

Sorry, Bullet has all of those hard coded internally. Nothing I could do right now. But see my request [here](https://github.com/bulletphysics/bullet3/issues/34) and the current Bullet patch we are using [here](https://github.com/RobotLocomotion/bullet3/pull/1) which essentially overwrites the hard coded tolerances in [btGjkEpa2.cpp](https://github.com/bulletphysics/bullet3/compare/bulletphysics:master...RobotLocomotion:master).

---

*[drake/systems/plants/collision/test/model_test.cc, line 84 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2402#-KI8gOYU7ipwYVqmP46x:-KI9OIdbEtd497XLTpes:1957389746) ([raw file](https://github.com/robotlocomotion/drake/blob/7547a380d5ebb2c1d26fd50204f37116abe7ebff/drake/systems/plants/collision/test/model_test.cc#L84)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> I am going to reviewable-resolve myself out of this discussion.
> 
> However, just to make sure I understand: the tolerance 1e-6 above used for EXPECT_NEAR is an _absolute_ tolerance, and the Eigen 1e-12 is a _proportional_ tolerance, correct?  Some of the downthread comments make me think this isn't 100% clear in the discussions.
</details>
@sherm1: is not an exact comparison because `points[0].getNormal()` was internally computed by Bullet, is not an analytical value. 
@jwnimmer-tri, @liangfok: yes, `eigen_matrix_compare.h` provides a more exhaustive evaluation of why exactly caused the matrices to be different, like it'll tell you what element is different. Very good for debugging. `isApprox` won't tell you what elements in the matrices are not equal but for sure it'll fail if matrices are different size or not equal by `tolerance`. I am ok with using either one, not a big gain in this case anyway. 
The change I propose is to simply change an inherently `int` check by a floating precision one. In this very simple test with simple geometries both pass. That won't be the case with more complex geometries (I am working on a gtest right now)

---

*[drake/systems/plants/collision/test/model_test.cc, line 86 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2402#-KI8lLIPLK7-9aRrnvub:-KI9Q2ngfAi-n-3IOF4i:-1582932834) ([raw file](https://github.com/robotlocomotion/drake/blob/7547a380d5ebb2c1d26fd50204f37116abe7ebff/drake/systems/plants/collision/test/model_test.cc#L86)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> Same here -- if these aren't exact results something is wrong. Both 0 and 0.5 are exactly representable in binary floating point.
</details>
yes, but the result of `points[0].getPtA()` is computed by Bullet with its own internal (hard coded) tolerances and we should not expect exact values.

---

*[drake/systems/plants/collision/test/model_test.cc, line 98 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2402#-KI8laMitI3UovrSdZh8:-KI9SfBwXg-TCWk185Wk:-227346749) ([raw file](https://github.com/robotlocomotion/drake/blob/7547a380d5ebb2c1d26fd50204f37116abe7ebff/drake/systems/plants/collision/test/model_test.cc#L98)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> OK, these can be approximate (should be accurate to around 1e-14, so this provides two orders of magnitude of slop). But actually I don't understand why we're expecting such a good solution here if we were only expecting 1e-6 earlier (that's a million times sloppier than 1e-12).
</details>
I changed tolerance to be machine precision, which in this case passes the test and is consistent with previous expectations in the "exact" comparison version.

---

*[drake/systems/plants/collision/test/model_test.cc, line 107 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2402#-KI8mA6g_PWRO8CrVT_m:-KI9Ss0tluGp8gT3LfG4:-1860472520) ([raw file](https://github.com/robotlocomotion/drake/blob/7547a380d5ebb2c1d26fd50204f37116abe7ebff/drake/systems/plants/collision/test/model_test.cc#L107)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> Shouldn't these last three be exact?
</details>
I don't think so. Everything inside `points` was computed by Bullet with (as we now know) many of its internals being iterative to (as we now know) some predefined (hard coded) tolerance.

---

*[drake/systems/plants/collision/test/model_test.cc, line 93 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2402#-KI8qDt_ccoMoscF6El-:-KI9QhGRG22_M6sF9Mb1:1785900208) ([raw file](https://github.com/robotlocomotion/drake/blob/4f39597991986b22e089a970a69468385f21eaaf/drake/systems/plants/collision/test/model_test.cc#L93)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> Better not to have magic constants if possible. I fed this one to one of my long-time [favorite web sites](http://isc.carma.newcastle.edu.au) -- the "inverse symbolic calculator" (you give it the answer it magically tells you how it was computed!). This appears to be (3*sqrt(2)-1)/2.
</details>
Haha nice! it actually is that value. However I got a different expression from geometric considerations. I changed it to that.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2402)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 1 of 4 files reviewed at latest revision, 9 unresolved discussions.

---

*[drake/systems/plants/collision/test/model_test.cc, line 84 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2402#-KI8gOYU7ipwYVqmP46x:-KI9WBPoPzy-RSR6Thu_:-289750271) ([raw file](https://github.com/robotlocomotion/drake/blob/7547a380d5ebb2c1d26fd50204f37116abe7ebff/drake/systems/plants/collision/test/model_test.cc#L84)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> @sherm1: is not an exact comparison because `points[0].getNormal()` was internally computed by Bullet, is not an analytical value. 
> @jwnimmer-tri, @liangfok: yes, `eigen_matrix_compare.h` provides a more exhaustive evaluation of why exactly caused the matrices to be different, like it'll tell you what element is different. Very good for debugging. `isApprox` won't tell you what elements in the matrices are not equal but for sure it'll fail if matrices are different size or not equal by `tolerance`. I am ok with using either one, not a big gain in this case anyway. 
> The change I propose is to simply change an inherently `int` check by a floating precision one. In this very simple test with simple geometries both pass. That won't be the case with more complex geometries (I am working on a gtest right now)
</details>
OK sounds good. Marking myself as "satisfied."

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2402)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Platform <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> +1 for cleaning up this mess!
</details>


---

Reviewed 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[drake/systems/plants/collision/Model.h, line 171 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2402#-KI8zwsj4vEy8Qb9hTdP:-KI9UO0hYtfK7iTqZ-rt:1827914304) ([raw file](https://github.com/robotlocomotion/drake/blob/4f39597991986b22e089a970a69468385f21eaaf/drake/systems/plants/collision/Model.h#L171)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> Fixed
</details>
BTW I was wrong to tag this one; it didn't need any fix on this particular line number.  Sorry!

---

*[drake/systems/plants/collision/test/model_test.cc, line 84 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2402#-KI8gOYU7ipwYVqmP46x:-KI9WFVlvMYywabn8AOb:2121887560) ([raw file](https://github.com/robotlocomotion/drake/blob/7547a380d5ebb2c1d26fd50204f37116abe7ebff/drake/systems/plants/collision/test/model_test.cc#L84)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> OK sounds good. Marking myself as "satisfied."
</details>
I still think this discussion is missing the point of changing from absolute to relative tolerance for comparisons, but I will remain reviewable-satisfied and let the dynamics-feature-review people do as they wish.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2402#-:-KI9WcikG5_R_qvK12L-:528945684)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
FWIW, the one gTest I am writing right now shows a case where Bullet computes the distance with an error of `1.000000082740371e-09`. The case only involves a box vs. sphere test. It so happens that when they actually are in contact the precision is smaller.

amcastro-tri commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[drake/systems/plants/collision/Model.h, line 171 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2402#-KI8zwsj4vEy8Qb9hTdP:-KI9haOPcD7fYs_LOGaB:11147320) ([raw file](https://github.com/robotlocomotion/drake/blob/4f39597991986b22e089a970a69468385f21eaaf/drake/systems/plants/collision/Model.h#L171)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> BTW I was wrong to tag this one; it didn't need any fix on this particular line number.  Sorry!
</details>
That's ok. I understood it as referencing the .cc file.

---

*[drake/systems/plants/collision/test/model_test.cc, line 84 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2402#-KI8gOYU7ipwYVqmP46x:-KI9hz3qp68hZ7tJDFOU:1362131522) ([raw file](https://github.com/robotlocomotion/drake/blob/7547a380d5ebb2c1d26fd50204f37116abe7ebff/drake/systems/plants/collision/test/model_test.cc#L84)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> I still think this discussion is missing the point of changing from absolute to relative tolerance for comparisons, but I will remain reviewable-satisfied and let the dynamics-feature-review people do as they wish.
</details>
for these simple tests I am satisfied with the absolute version. But open to change if someone is strong for another option. I just like that `isApprox` is a short call, that gets the job done and, is an Eigen builtin functionality. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2402)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[drake/systems/plants/collision/test/model_test.cc, line 84 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2402#-KI8gOYU7ipwYVqmP46x:-KI9jLIgK4SKxa9ScLYh:-929433963) ([raw file](https://github.com/robotlocomotion/drake/blob/7547a380d5ebb2c1d26fd50204f37116abe7ebff/drake/systems/plants/collision/test/model_test.cc#L84)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> for these simple tests I am satisfied with the absolute version. But open to change if someone is strong for another option. I just like that `isApprox` is a short call, that gets the job done and, is an Eigen builtin functionality. 
</details>
To be super clear: `isApprox` is relative (proportional) tolerance, and `EXPECT_NEAR` is absolute tolerance.  (You say "satisfied with absolute" and "like isApprox" which makes me think I've not been clear yet; `isApprox` is not absolute.)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2402)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[drake/systems/plants/collision/test/model_test.cc, line 84 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2402#-KI8gOYU7ipwYVqmP46x:-KI9rA6IkACQNwNWYynt:-1501582717) ([raw file](https://github.com/robotlocomotion/drake/blob/7547a380d5ebb2c1d26fd50204f37116abe7ebff/drake/systems/plants/collision/test/model_test.cc#L84)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> To be super clear: `isApprox` is relative (proportional) tolerance, and `EXPECT_NEAR` is absolute tolerance.  (You say "satisfied with absolute" and "like isApprox" which makes me think I've not been clear yet; `isApprox` is not absolute.)
</details>
Very true! I should read the documentation. In any case, satisfied with `isApprox`. Notice the things I'm comparing are O(1) so it really does not make much of a difference here. I personally feel inclined to use built in functionality. But again, I can change it. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2402)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
> when they actually are in contact the precision is smaller

"precision is smaller" is ambiguous -- it can mean "less precise" or "the precision is a smaller number like 1e-12, so more precise". To avoid this problem, I try to use "loose" and "tight" ("looser" and "tighter") when I talk about precision and accuracy -- I recommend that terminology because it will save you a lot of explanation time!

Feature <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>



---

Reviewed 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/systems/plants/collision/test/model_test.cc, line 84 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2402#-KI8gOYU7ipwYVqmP46x:-KI9xtn_P52ViOfovyW-:-1611799358) ([raw file](https://github.com/robotlocomotion/drake/blob/7547a380d5ebb2c1d26fd50204f37116abe7ebff/drake/systems/plants/collision/test/model_test.cc#L84)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> Very true! I should read the documentation. In any case, satisfied with `isApprox`. Notice the things I'm comparing are O(1) so it really does not make much of a difference here. I personally feel inclined to use built in functionality. But again, I can change it. 
</details>
I'm also satisfied but would like to state that absolute tolerance should not be used for floating point comparisons except in very narrow, well documented circumstances. Absolute tolerance is worse than useless in most cases. Being near "1" though is one of the cases where it doesn't matter! It is nice that Eigen::isApprox() uses relative tolerance but that actually isn't sufficient either because it should degrade to absolute tolerance when the numbers get small enough.

---

*[drake/systems/plants/collision/test/model_test.cc, line 93 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2402#-KI8qDt_ccoMoscF6El-:-KI9zt_bn82SIETPBw4C:-400587012) ([raw file](https://github.com/robotlocomotion/drake/blob/4f39597991986b22e089a970a69468385f21eaaf/drake/systems/plants/collision/test/model_test.cc#L93)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> Haha nice! it actually is that value. However I got a different expression from geometric considerations. I changed it to that.
</details>
Great, thanks!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2402#-:-KI9wHgUUf6-w-GvzBrd:-1000782066)*
<!-- Sent from Reviewable.io -->
